src/sqlfluff/cli/commands.py

```diff
@@ -476,6 +477,16 @@ def fix(force, paths, processes, bench=False, fixed_suffix="", logger=None, **kw
     if fixing_stdin:
         stdin = sys.stdin.read()
         result = lnt.lint_string_wrapped(stdin, fname="stdin", fix=True)
+        # exit early if there is a templater error.
+        if any((isinstance(i, SQLTemplaterError) for i in result.get_violations())):
+            click.echo(
+                colorize(
+                    "The stdin input contains template variables that could not be "
+                    + "parsed. Aborting fix.",
+                    "red",
+                )
+            )
+            sys.exit(1)
         stdout = result.paths[0].files[0].fix_string()[0]
```

Comment:
When you test it using files it doesn't "fix" the file either. It just doesn't bomb out like it does with the CLI. So switching to linting a temp file won't help any more than the current fix.

Digging into why this is, it is because it does this for the file option:

```py
    # NB: We filter to linting violations here, because they're
    # the only ones which can be potentially fixed.
    if result.num_violations(types=SQLLintError, fixable=True) > 0:
        click.echo("==== fixing violations ====")
        ...
```

So it only attempts to fix violations if there are any fixable violations! Whereas the stdin method attempts to fix regardless.

Therefore I think you should implement this same thing instead of checking for just `SQLTemplaterError` errors:

```suggestion
    # handle stdin case. should output formatted sql to stdout and nothing else.
    if fixing_stdin:
        stdin = sys.stdin.read()
        result = lnt.lint_string_wrapped(stdin, fname="stdin", fix=True)
        if result.num_violations(types=SQLLintError, fixable=True) > 0:
            stdout = result.paths[0].files[0].fix_string()[0]
        else:
            if verbose:
                click.echo("No fixable violations found.")
            stdout = stdin
```

Note that I'm suggesting returning `stdin` as `stdout`, which is what happens in non-templated SQL that doesn't need fixing, but that only happens because `result.paths[0].files[0].fix_string()[0]` is set to the same as input, whereas that doesn't happen on templating failures, but net effect is the same. So might as well be explicit about using stdin.

Reply:
That sounds great to me!

Reply:
Waiiiiit, does this error only raise when there are no fixable errors? I ask because:

```
➜ echo 'create TABLE {{ params.dsfsdfds }}.t (a int)' | sqlfluff fix - --verbose
The stdin input contains template variables that could not be parsed. Aborting fix.

➜ echo 'create TABLE a.t (a int)' | sqlfluff fix - --verbose
create table a.t (a int)
```

Both cases have a fixable error and in one we ALSO have a template error which could (?) be handled and a fix returned. So is this now a two-case issue?

1. if no fixable errors, return stdin even if there was an exception on the way (duh)
2. Use a tempfile approach to support fixing what can be fixed using a single code path

am i understanding the suggestion correctly?

Reply:
FYI- with this implementation we still raised a templated file error

```py
    if fixing_stdin:
        stdin = sys.stdin.read()
        with tempfile.NamedTemporaryFile(mode="w+", suffix=".sql") as f:
            f.write(stdin)
            f.flush()
            result = lnt.lint_path(f.name, fix=True)
            stdout = result.files[0].fix_string()[0]
        click.echo(stdout, nl=False)
        sys.exit()
```


```
squidward 🐙 in sqlfluff on  nolan/fix-stdout-error [!] via sqlfluff via 🐍 sqlfluff 
➜ echo 'create TABLE a.t (a int)' | sqlfluff fix - --verbose                    
create table a.t (a int)

squidward 🐙 in sqlfluff on  nolan/fix-stdout-error [!] via sqlfluff via 🐍 sqlfluff 
➜ echo 'create TABLE {{ params.dsfsdfds }}.t (a int)' | sqlfluff fix - --verbose
Traceback (most recent call last):
  File "/home/squidward/.pyenv/versions/sqlfluff/bin/sqlfluff", line 33, in <module>
    sys.exit(load_entry_point('sqlfluff', 'console_scripts', 'sqlfluff')())
  File "/home/squidward/.pyenv/versions/sqlfluff/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/home/squidward/.pyenv/versions/sqlfluff/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/squidward/.pyenv/versions/sqlfluff/lib/python3.8/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/squidward/.pyenv/versions/sqlfluff/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/squidward/.pyenv/versions/sqlfluff/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/squidward/Documents/repos/sqlfluff/src/sqlfluff/cli/commands.py", line 483, in fix
    stdout = result.files[0].fix_string()[0]
  File "/home/squidward/Documents/repos/sqlfluff/src/sqlfluff/core/linter/linted_file.py", line 242, in fix_string
    linter_logger.debug("Original Tree: %r", self.templated_file.templated_str)
AttributeError: 'NoneType' object has no attribute 'templated_str'

Reply:
> Waiiiiit, does this error only raise when there are no fixable errors? I ask because:

No. It only raises when there is a templating error. As in that case the `templated_file` is not set and that is needed by `fix_string()`.

For non-templates, no fixable errors that variable is set - but to the original SQL anyway. BUT for case the fixed is not outputted in file mode (try it with a file and the SUFFIX option and you’ll see it’s not written).

> Both cases have a fixable error and in one we ALSO have a template error

That is not possible. Template errors result in unparsable SQL and hence no fixable errors. Which may seem a little harsh but is actually right IMHO. If you don’t understand the templates pieces then you don’t understand the SQL. For simple cases you could argue to just ignore the templates but think that’s a risky option. To be honest it’s the same with unparseable sections but I think we do allow them to be fixed (but not sure we should).

Reply:
> FYI- with this implementation we still raised a templated file error

Are you sure? Not at computer now but can look later but was definitely working earlier. Maybe I copied and pasted the fix in, incorrectly.

Reply:
Ohhhh i see- the parser will be unable to find any fixable violations whenever the templater fails. That makes sense and nooooowwww i understand the suggestion. In that case we have two choices:

1. Spit the stdin back out or not. It feels good to do so from a unix/posix sort of perspective
2. exit zero or nonzero, optionally logging an informative message

I lean towards exiting nonzero AND spitting the stdin out. users should be _somehow_ notified that something is not right

Reply:
Yup works for me. But think that should be done for ANY non-fixable errors not just template errors?

But that might cause other breakages?

BTW if we ever did start supporting fixing SQL with template errors (which as I say I’m not sure we should, but anyway) then my suggestion would seem less let support that (as fixable errors would now be > 0).

Reply:
Yeah aboslutely; i think exit nonzero if any templater error but not if the sql is perfect. will add tests to make sure all is good.

Reply:
I don't know the details, but some of the lint rules _will_ run on unparseable code. Search the rule code for `_works_on_unparsable`. Some rules specify `False`, but not all.

Reply:
Linting is fine. Fixing is more questionable.

But maybe OK for simple fixes (e.g. capitalisation, some spacing).

Anyway going off track.

BTW I do wonder if `fix_string()` should be made more robust rather than failing so ungracefully when `templated_file` doesn’t exist? Yes we probably shouldn’t waste our time fixing SQL with no fixable errors but still…

Reply:
My take is that the use of fix_string here is something of a holdover that we never re-wrote to be more robust. It goes back to like my first fluff pr i think. @barrywhart is right and we should seek to unify the code path between stdin and file fixing; probably a tempfile solution would work.

Reply:
Well I think they are unified now. We only call fix if there are fixable errors. That’s what the difference was and that is now fixed.